enhancements/aiops/20230203-anomaly-keyword-filter.md

```diff
@@ -0,0 +1,79 @@
+# Title: 
+Filter out anomalous keywords from training dataset of workload logs for Deep Learning model
+
+## Summary: 
+Currently, when a user would like to train a Deep Learning model on a watchlist of workloads, the corresponding logs for all of the workloads specified are fetched from Opensearch within the last hour. While a check is made within the Opensearch query to omit any logs which were previously marked as anomalous, there is no check made for any workload logs which contain keywords that are typically associated with anomalous logs. I propose that we maintain a list of anomalous keywords and for any log message that is fetched from Opensearch, we do not include it in the training dataset if it contains at least one word from the list of anomalous keywords.
+
+## Use case: 
+This will filter out workload log messages with anomalous keywords from the training data of the Deep Learning model. 
+
+## Benefits: 
+* Acts as safe guard to avoid adding clearly anomalous log messages to training dataset.
+* Improves insights provided by Deep Learning model.
+
+
+## Impact: 
+There would not be any major impact to the current system. The query made to Opensearch to fetch the filtered log messages will need to be updated but no architecture change is necessary.
+
+## Implementation details: 
+Error keywords:
+- fail
+- error
+- fatal
+- exception
+- timeout
+- unavailable
+- OOM
+- crash
+- connection refused
+- network error
+- deadlock
+- disk space
+- out of disk
+- high load
+
+The GPU controller service, which handles both model training and inferencing, will query OpenSearch to retrieve workload logs from the past hour that have not been marked as anomalous and do not contain any of the designated anomalous keywords. The retrieval process will be performed using scrolling, where each query to OpenSearch will retrieve 10,000 logs at a time. This is done in preparation for training a Deep Learning model on a designated watchlist of workload logs.
+
+After retrieving all log messages for the training dataset, the GPU controller will proceed to create the required model vocabulary and commence training the model, using the same steps as previously described.
+
+
+## Acceptance criteria: 
+* Deep Learning models should only be trained on workload logs within the last hour that are not marked as anomalous and do not contain any of the designated anomalous keywords.
```

Comment:
Is this the last hour relative to real time? Or the last hour worth of logs with respect to the deployment in opensearch?

Reply:
Last hour relative to real time.